### PR TITLE
implement deadlock detection

### PR DESCRIPTION
**Pull request by reAsOn2010**
WIP for #774 

I'm using pjdfstest to verify, testcases used is listed below
```
$ cat runtest/syscalls-ddlk
fcntl17 fcntl17
fcntl17_64 fcntl17_64
```

Main branch:
```
$ ./runltp -d /mnt/jfs -f syscalls-ddlk
<some log>
Running tests.......
fcntl17     0  TINFO  :  Enter preparation phase
fcntl17     0  TINFO  :  Exit preparation phase
fcntl17     0  TINFO  :  Enter block 1
fcntl17     0  TINFO  :  child 1 starting
fcntl17     0  TINFO  :  child 1 pid 82931 locked
fcntl17     0  TINFO  :  child 2 starting
fcntl17     0  TINFO  :  child 2 pid 82932 locked
fcntl17     0  TINFO  :  child 3 starting
fcntl17     0  TINFO  :  child 3 pid 82933 locked
fcntl17     0  TINFO  :  child 3 resuming
fcntl17     0  TINFO  :  child 2 resuming
fcntl17     0  TINFO  :  child 1 resuming
fcntl17     0  TINFO  :  child 1 unlocked
fcntl17     1  TFAIL  :  fcntl17.c:429: Alarm expired, deadlock not detected
fcntl17     0  TWARN  :  fcntl17.c:430: You may need to kill child processes by hand
fcntl17     2  TFAIL  :  fcntl17.c:408: Unexpected death of child process
fcntl17     0  TWARN  :  tst_tmpdir.c:336: tst_rmdir: rmobj(/mnt/jfs/ltp-6SyZCZsNEH/fcndBzo6b) failed: unlink(/mnt/jfs/ltp-6SyZCZsNEH/fcndBzo6b) failed; errno=2: ENOENT
fcntl17     3  TFAIL  :  fcntl17.c:363: parent_wait() failed
fcntl17     4  TFAIL  :  fcntl17.c:605: child 2 didn't deadlock, returned: 4
fcntl17     5  TFAIL  :  fcntl17.c:295: fcntl on file failed, errno =9
fcntl17     6  TFAIL  :  fcntl17.c:295: fcntl on file failed, errno =9
fcntl17     7  TBROK  :  tst_sig.c:232: unexpected signal SIGPIPE(13) received (pid = 82930).
fcntl17     8  TBROK  :  tst_sig.c:232: Remaining cases broken
fcntl17     0  TINFO  :  Enter preparation phase
fcntl17     0  TINFO  :  Exit preparation phase
fcntl17     0  TINFO  :  Enter block 1
fcntl17     0  TINFO  :  child 1 starting
fcntl17     0  TINFO  :  child 1 pid 82946 locked
fcntl17     0  TINFO  :  child 2 starting
fcntl17     0  TINFO  :  child 2 pid 82947 locked
fcntl17     0  TINFO  :  child 3 starting
fcntl17     0  TINFO  :  child 3 pid 82948 locked
fcntl17     0  TINFO  :  child 3 resuming
fcntl17     0  TINFO  :  child 1 resuming
fcntl17     0  TINFO  :  child 2 resuming
fcntl17     0  TINFO  :  child 1 unlocked
fcntl17     1  TFAIL  :  fcntl17.c:429: Alarm expired, deadlock not detected
fcntl17     0  TWARN  :  fcntl17.c:430: You may need to kill child processes by hand
fcntl17     2  TFAIL  :  fcntl17.c:408: Unexpected death of child process
fcntl17     0  TWARN  :  tst_tmpdir.c:336: tst_rmdir: rmobj(/mnt/jfs/ltp-6SyZCZsNEH/fcnYP2Nas) failed: unlink(/mnt/jfs/ltp-6SyZCZsNEH/fcnYP2Nas) failed; errno=2: ENOENT
fcntl17     3  TFAIL  :  fcntl17.c:363: parent_wait() failed
fcntl17     4  TFAIL  :  fcntl17.c:605: child 2 didn't deadlock, returned: 4
fcntl17     5  TFAIL  :  fcntl17.c:295: fcntl on file failed, errno =9
fcntl17     6  TFAIL  :  fcntl17.c:295: fcntl on file failed, errno =9
fcntl17     7  TBROK  :  tst_sig.c:232: unexpected signal SIGPIPE(13) received (pid = 82945).
fcntl17     8  TBROK  :  tst_sig.c:232: Remaining cases broken
INFO: ltp-pan reported some tests FAIL
<some log>

$ cat /opt/ltp/results/LTP_RUN_ON-2024_03_08-10h_31m_29s.log
Test Start Time: Fri Mar  8 10:31:30 2024
-----------------------------------------
Testcase                                           Result     Exit Value
--------                                           ------     ----------
fcntl17                                            FAIL       7
fcntl17_64                                         FAIL       7

-----------------------------------------------
Total Tests: 2
Total Skipped Tests: 0
Total Failures: 2
Kernel Version: 6.1.79
Machine Architecture: x86_64
Hostname: pat
```

This branch:
```
$ ./runltp -d /mnt/jfs -f syscalls-ddlk
<some log>
Running tests.......
<<<test_start>>>
tag=fcntl17 stime=1709865314
cmdline="fcntl17"
contacts=""
analysis=exit
<<<test_output>>>
fcntl17     0  TINFO  :  Enter preparation phase
fcntl17     0  TINFO  :  Exit preparation phase
fcntl17     0  TINFO  :  Enter block 1
fcntl17     0  TINFO  :  child 1 starting
fcntl17     0  TINFO  :  child 1 pid 84263 locked
fcntl17     0  TINFO  :  child 2 starting
fcntl17     0  TINFO  :  child 2 pid 84264 locked
fcntl17     0  TINFO  :  child 3 starting
fcntl17     0  TINFO  :  child 3 pid 84265 locked
fcntl17     0  TINFO  :  child 2 resuming
fcntl17     0  TINFO  :  child 3 resuming
fcntl17     0  TINFO  :  child 1 resuming
fcntl17     0  TINFO  :  child 1 unlocked
fcntl17     0  TINFO  :  child 3 lockw err 35
fcntl17     0  TINFO  :  child 1 exiting
fcntl17     0  TINFO  :  child 3 exiting
fcntl17     0  TINFO  :  child 2 lockw err 35
fcntl17     0  TINFO  :  child 2 exiting
fcntl17     1  TPASS  :  Block 1 PASSED
fcntl17     0  TINFO  :  Exit block 1
<<<execution_status>>>
initiation_status="ok"
duration=0 termination_type=exited termination_id=0 corefile=no
cutime=0 cstime=0
<<<test_end>>>
<<<test_start>>>
tag=fcntl17_64 stime=1709865314
cmdline="fcntl17_64"
contacts=""
analysis=exit
<<<test_output>>>
incrementing stop
fcntl17     0  TINFO  :  Enter preparation phase
fcntl17     0  TINFO  :  Exit preparation phase
fcntl17     0  TINFO  :  Enter block 1
fcntl17     0  TINFO  :  child 1 starting
fcntl17     0  TINFO  :  child 1 pid 84267 locked
fcntl17     0  TINFO  :  child 2 starting
fcntl17     0  TINFO  :  child 2 pid 84268 locked
fcntl17     0  TINFO  :  child 3 starting
fcntl17     0  TINFO  :  child 3 pid 84269 locked
fcntl17     0  TINFO  :  child 2 resuming
fcntl17     0  TINFO  :  child 3 resuming
fcntl17     0  TINFO  :  child 1 resuming
fcntl17     0  TINFO  :  child 3 lockw err 35
fcntl17     0  TINFO  :  child 3 exiting
fcntl17     0  TINFO  :  child 1 unlocked
fcntl17     0  TINFO  :  child 1 exiting
fcntl17     0  TINFO  :  child 2 lockw err 35
fcntl17     0  TINFO  :  child 2 exiting
fcntl17     1  TPASS  :  Block 1 PASSED
fcntl17     0  TINFO  :  Exit block 1
<<<execution_status>>>
initiation_status="ok"
duration=0 termination_type=exited termination_id=0 corefile=no
cutime=0 cstime=0
<<<test_end>>>
<some log>

$ cat /opt/ltp/results/LTP_RUN_ON-2024_03_08-10h_35m_13s.log
Test Start Time: Fri Mar  8 10:35:14 2024
-----------------------------------------
Testcase                                           Result     Exit Value
--------                                           ------     ----------
fcntl17                                            PASS       0
fcntl17_64                                         PASS       0

-----------------------------------------------
Total Tests: 2
Total Skipped Tests: 0
Total Failures: 0
Kernel Version: 6.1.79
Machine Architecture: x86_64
Hostname: pat
```

Some advice needed about this PR:
1. I don't know if it will have much performance impact introduced by the newly added detection. So I copied the MAX ITERATION 10 from the kernel.
2. A new Redis hash set introduced here. But simple key/value  pair is also enough to use here. Which one is preferred?
3. If the data remained in meta store in any case, currently no clean up will be done.

If this implementation is good, I'll do the same on other meta store type.

### CODE REVIEW

**Comment by reAsOn2010** [comment]
Oh, seems that some unit tests needed. In pkg/meta/base_test.go?

**Comment by davies** [comment]
> Oh, seems that some unit tests needed. In pkg/meta/base_test.go?

yes

File: pkg/meta/redis_lock.go [hunk 1]
```diff
@@ -175,15 +175,29 @@ func (r *redisMeta) Setlk(ctx Context, inode Ino, owner uint64, block bool, ltyp
...
 				for _, l := range ls {
 					// find conflicted locks
 					if (ltype == F_WRLCK || l.Type == F_WRLCK) && end >= l.Start && start <= l.End {
-						return syscall.EAGAIN
+						conflictOwner = owner
+						err = syscall.EAGAIN
 					}
 				}
 			}
+			if block && err == syscall.EAGAIN {
```
Comment (davies) [inline pkg/meta/redis_lock.go:189]: We can delay the detect after waiting for 1 second

File: pkg/meta/redis_lock.go [hunk 2]
```diff
@@ -175,15 +175,29 @@ func (r *redisMeta) Setlk(ctx Context, inode Ino, owner uint64, block bool, ltyp
...
+				tx.HSet(ctx, r.conflictKey(), lkey, conflictOwner)
```
Comment (davies) [inline pkg/meta/redis_lock.go:190]: cleanup it in the end

**Comment by reAsOn2010** [comment]
@davies 
Because I need to wait for some sleep loop, so the detection now implemented in a new transaction. 
A simple test is implemented.

**Comment by davies** [comment]
#774 is about POSIX lock, but this PR is dealing with flock, they are different. POSIX lock will be difficult to find dead lock.

Based on man page of flock, linux does NOT detect deadlock.

Since PR is not ready yet, we can close it for now. It's better to work on POSIX locks if you want to continue.

**Pull request closed without merging**